### PR TITLE
Add: tab audible behaviour

### LINKED ISSUE #7: Lower music volume when Chrome plays sound
Whenever something in Chrome plays sound (a video, a message notification, etc,) it'd be great if the extension's volume would automatically lower.

### PR DESCRIPTION
**Pull request by PikaDude**
This PR adds a customisable function that will trigger whenever a tab begins to play audio, such as lowering the extension's music volume or pausing it. Also includes some other small changes and fixes.

Closes animal-crossing-music-extension/Animal-Crossing-Music-Extension#7
image

### CODE REVIEW

**Comment by jordigarcl** [comment]
Will check this now

File: scripts/background/TabAudioHandler.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        });
+    }
+
+    function removeHandler() {
+        chrome.tabs.onUpdated.removeListener(tabUpdatedHandler);
+        tabUpdatedHandler = null;
+    }
+
+    function checkTabs() {
+        // A tab can be muted and still be "audible"
```
Comment (jordigarcl) [inline scripts/background/TabAudioHandler.js:39]: I'm curious, how is that so?

File: scripts/background/AudioManager.js [hunk 2]
```diff
@@ -164,6 +169,54 @@ function AudioManager(addEventListener, isTownTune) {
...
+			if (audio.paused && tabAudio != 'pause') {
+				audio.play();
+				tabAudioPaused = false;
+				window.notify("unpause");
+			}
+			if (reducedVolume && tabAudio != 'reduce') {
+				reducedVolume = false;
+				audio.volume = setVolume;
+			}
+			window.notify("tabAudio", [true, tabAudio, reduceValue]);
```
Comment (jordigarcl) [inline scripts/background/AudioManager.js]: Current logic introduces following bug:
If music **from the extension** is playing (and no other tabs are playing music), when a 'Tab Audio Behaviour' option is switched to 'reduce' /'pause', extension music is reduced/paused. 

That is, switching to 'pause' will effectively pause the extension even though there are no other sounds.

**Comment by jordigarcl** [comment]
Wow, great work.

Although I think the nested `if`s on the AudioManager's `tabAudio` event listener could be refactored (maybe leveraging JS's switch statement), code looks good!

I did found a bug that occurs when switching options with the extension enabled and no external music playing. As it is not a critical bug (and is trivial to fix) I'm unblocking the PR!  👍

**Comment by jordigarcl** [comment]
Wait, just found a more important bug which should be  strongly related to the first one.
For some reason manually pausing extension music and then triggering the `chrome.tabs.onUpdated` callback (for example by reloading or creating/deleting a tab) makes the extension play music automatically. This only happens when the 'pause music' option is selected. Banner is (correctly) not updated.

**Review by jordigarcl — APPROVED** [review]
I think the 'music re-starting automatically when updating tabs' is critical enough to be addressed before merging

**Review by jordigarcl — CHANGES_REQUESTED** [review]
Ooops

**Comment by PikaDude** [comment]
The `tabAudio` event listener is purposely left as it is to make it a bit more human-readable and easier to follow. I found that refactoring it made it hard to read and follow. Also nice catch on those bugs! They were all related to a small mistake that I've now fixed. I also noticed and fixed another issue that was causing this function to try and pause the music while it was still loading, throwing errors that didn't really mean anything, but errors nonetheless.

**Review by jordigarcl — APPROVED** [review]
Last change fixed the music stopping bug 🥳

**Pull request merged**